docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+following the same design principles, this includes functionality that interacts with specific non-stable versions of
+third party dependencies such as Tendermint
+* the core API doesn't implement *any* functionality, it just defines types
+* go stable API compatibility guidelines are followed: https://go.dev/blog/module-compatibility
+
+A "runtime" module is any module which implements the core functionality of composing an ABCI app, which is currently
+handled by `BaseApp` and the `ModuleManager`. Runtime modules which implement the core API are *intentionally* separate
+from the core API in order to enable more parallel versions and forks of the runtime module than is possible with the
+SDK's current tightly coupled `BaseApp` design while still allowing for a high degree of composability and
+compatibility.
+
+Modules which are built only against the core API don't need to know anything about which version of runtime,
+`BaseApp` or Tendermint in order to be compatible. Modules from the core mainline SDK could be easily composed
+with a forked version of runtime with this pattern.
+
+This design is intended to enable matrices of compatible dependency versions. Ideally a given version of any module
+is compatible with multiple versions of the runtime module and other compatible modules. This will allow dependencies
+to be selectively updated based on battle-testing. More conservative projects may want to update some dependencies
+slower than more fast moving projects.
+
+### Core Services
+
+The following "core services" are defined by the core API. All valid runtime module implementations should provide
+implementations of these services to modules via both [dependency injection](./adr-057-app-wiring-1.md) and
+manual wiring. The individual services described below are all bundled in a convenient `appmodule.Service`
+"bundle service" so that for simplicity modules can declare a dependency on a single service.
+
+#### Store Services
+
+Store services will be defined in the `cosmossdk.io/core/store` package.
+
+The generic `store.KVStore` interface is the same as current SDK `KVStore` interface. Store keys have been refactored
+into store services which, instead of expecting the context to know about stores, invert the pattern and allow
+retrieving a store from a generic context. There are three store services for the three types of currently supported
+stores - regular kv-store, memory, and transient:
+
+```go
+type KVStoreService interface {
+    OpenKVStore(context.Context) KVStore
+}
+
+type MemoryStoreService interface {
+    OpenMemoryStore(context.Context) KVStore
+}
+type TransientStoreService interface {
+    OpenTransientStore(context.Context) KVStore
+}
+```
+
+Modules can use these services like this:
+```go
+func (k msgServer) Send(ctx context.Context, msg *types.MsgSend) (*types.MsgSendResponse, error) {
+	store := k.kvStoreSvc.OpenKVStore(ctx)
```

Comment:
related to the above comment: it seems that we always do `k.{someService}.{whatWeActuallyWant}`. Most services only have 1 method. This new `.Open` seems unnecessary here.

For some of these core services: can we simplify it to simply `k.{whatWeActuallyWant}`.

```diff
type keeper struct {
	depinject.In

-	KVStoreKey       store.KVStoreService
+	KVStoreKey       store.KVStore
	BlockInfoService blockinfo.Service
-	EventService     event.Service
+	EventMgr         event.Manager
	GasService       gas.Service
}
```

Reply:
We still need to interact with the context to get a valid instance. Maybe the bundle service simplifies this?

Reply:
Ah I see. I remember a discussion about adding the `context.Context` argument to all Get/Set/Delete/etc methods. Maybe Open is better.

Reply:
If there is more than one store operation in a method than open will be shorter